### PR TITLE
Feature/chapter 2

### PR DESCRIPTION
**Pull request by JBartscher**
Implement bats test for chapter 2 (services)

### CODE REVIEW

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		#It is recommended to use only one tool for time synchronization, be it ntp, chrony or the systemd-timesyncd service.
+		#If ntp is the chosen tool, this test verifies that ntp is configured correctly.
+
+		run bash -c "dpkg -s chrony"
+		[[ "${lines[0]}" == "dpkg-query: package 'chrony' is not installed and no information is available" ]]
+
+		run bash -c "systemctl is-enabled systemd-timesyncd.service"
+		[[ "$output" = "masked" ]]
+
+		run bash -c "grep "^restrict" /etc/ntp.conf"
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats]: I think this needs some escaping:
```suggestion
		run bash -c "grep \"^restrict\" /etc/ntp.conf"
```

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		[[ "$output" == *"restrict -6 default "*"kod"* ]]
+		[[ "$output" == *"restrict -6 default "*"nomodify"* ]]
+		[[ "$output" == *"restrict -6 default "*"notrap"* ]]
+		[[ "$output" == *"restrict -6 default "*"nopeer"* ]]
+		[[ "$output" == *"restrict -6 default "*"noquery"* ]]
+		# 2.2.1.4 may fail due to "-4" or "-6" not being present in conf file
+		run bash -c "grep -E \"^(server|pool)\" /etc/ntp.conf"
+		[ "$status" -eq 0 ]
+		[[ "$output" == "server "* ]] || [[ "$output" == "pool "* ]]
+		run bash -c "grep "RUNASUSER=ntp" /etc/init.d/ntp"
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats]:
```suggestion
		run bash -c "grep \"RUNASUSER=ntp\" /etc/init.d/ntp"
```

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		[[ "${lines[5]}" == *"NTP service: active"* ]]
+}
+
+@test "2.1.1.3 Ensure chrony is configured (Automated)" {
+		#It is recommended to use only one tool for time synchronization, be it ntp, chrony or the systemd-timesyncd service.
+		#If chrony is the chosen tool, this test verifies that chrony is configured correctly
+
+		run bash -c "dpkg -s ntp | grep -E '(Status:|not installed)'"
+		[[ "${lines[0]}" == *"dpkg-query: package 'ntp' is not installed and no information is available"* ]]
+
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats:33]: The check for systemd-timesyncd being masked is missing here

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		run bash -c "grep -E \"^(server|pool)\" /etc/ntp.conf"
+		[ "$status" -eq 0 ]
+		[[ "$output" == "server "* ]] || [[ "$output" == "pool "* ]]
+		run bash -c "grep "RUNASUSER=ntp" /etc/init.d/ntp"
+		[ "$status" -eq 0 ]
+		[[ "$output" == "RUNASUSER=ntp" ]]
+}
+
+@test "2.1.2 Ensure X Window System is not installed (Automated)" {
+			run bash -c "dpkg -l xserver-xorg*"
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats]:
```suggestion
	run bash -c "dpkg -l xserver-xorg*"
```

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+		[ "$status" -eq 1 ]
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats:79]:
```suggestion
	[ "$status" -eq 1 ]
```

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		[[ "$output" == *"no packages found matching xserver-xorg*" ]]
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats:80]:
```suggestion
	[[ "$output" == *"no packages found matching xserver-xorg*" ]]
```

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+@test "2.1.3 Ensure Avahi Server is not installed (Automated)" {
+		run bash -c "systemctl is-enabled avahi-daemon"
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats]: This is not the command from the CIS document

File: 2-Services/2.1-Special-Purpose-Services.bats [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		if [ "$status" -eq 0 ]; then
+				[[ "$output" != "enabled" ]]
+		fi
+}
+
+@test "2.1.4 Ensure CUPS is not installed (Automated)" {
+      run bash -c "systemctl is-enabled cups"
```
Comment (robertauer) [inline 2-Services/2.1-Special-Purpose-Services.bats]: This is not the command from the CIS document

File: 2-Services/2.3-ensure-nonessential-services-are-removed-or-masked.bats [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bats
+
+@test "2.3 Ensure nonessential services are removed or masked (Manual)" {
+    skip "This audit has to be done manually"
+
+    # Run the following command and review ports/network-configuration:
```
Comment (robertauer) [inline 2-Services/2.3-ensure-nonessential-services-are-removed-or-masked.bats]:
```suggestion
    # Run the following audit command:
```

File: 2-Services/2.3-ensure-nonessential-services-are-removed-or-masked.bats [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+    # lsof -i -P -n | grep -v "(ESTABLISHED)"
```
Comment (robertauer) [inline 2-Services/2.3-ensure-nonessential-services-are-removed-or-masked.bats:7]: Maybe we can just use the advise text from the CIS document
```suggestion
    # lsof -i -P -n | grep -v "(ESTABLISHED)"
    # Review the output to ensure that all services listed are required on the system. If a listed
    # service is not required, remove the package containing the service. If the package
    # containing a non-essential service is required, stop and mask the non-essential service.
```

**Review by robertauer — CHANGES_REQUESTED** [review]

File: 2-Services/2.1-Special-Purpose-Services.bats (hunk 1 above)
Comment (JBartscher) [reply 2-Services/2.1-Special-Purpose-Services.bats]: Good find

File: 2-Services/2.3-ensure-nonessential-services-are-removed-or-masked.bats (hunk 10 above)
Comment (JBartscher) [reply 2-Services/2.3-ensure-nonessential-services-are-removed-or-masked.bats:7]: I do not have any preferences, I will implement these changes

File: 2-Services/2.1-Special-Purpose-Services.bats (hunk 3 above)
Comment (JBartscher) [reply 2-Services/2.1-Special-Purpose-Services.bats:33]: If I understand correctly, you are referring to the following sentence: 
`Note: If chrony or ntp are used, systemd-timesyncd should be stopped and masked, and  this section skipped`.
I would add a test step in the steps that check the configuration of chrony and ntp to see if systemd-timesyncd is correctly set to "masked".

Comment (robertauer) [reply 2-Services/2.1-Special-Purpose-Services.bats:33]: There is one already defined in the CIS document.
```
Run the following command to verify that systemd-timsyncd is masked:
# systemctl is-enabled systemd-timesyncd
masked
```
And you already have a test in the 2.1.1.4 section for this:
```
run bash -c "systemctl is-enabled systemd-timesyncd.service"
[[ "$output" = "masked" ]]
```

**Review by robertauer — APPROVED** [review]

**Pull request merged**